### PR TITLE
Tap mod swap magic

### PR DESCRIPTION
**Pull request by osamuaoki**
This is for  https://github.com/qmk/qmk_firmware/issues/18349

This worked for my case of combination of
 * mod-tap MT(MOD_LCTL, KC_ESC)
 * MAGIC MAGIC_SWAP_ESCAPE_CAPS

At least some effort is made to avoid applying swap for non-keycode
Other case are included and made to be synmetrical.
Code may need some reformatting.  (/* ...*/ comment)
This is more-or-less proof-of-concept patch.
So there may be shortcomings.

## Description

See issue. https://github.com/qmk/qmk_firmware/issues/18349

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

This solves  https://github.com/qmk/qmk_firmware/issues/18349

## Checklist

- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [NO] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage). --- Yes tested but not exhaustive yet.

### CODE REVIEW

File: quantum/keycode_config.c [hunk 1]
```diff
@@ -24,94 +24,103 @@ extern keymap_config_t keymap_config;
  * and will return the corrected keycode, when appropriate.
  */
 uint16_t keycode_config(uint16_t keycode) {
-    switch (keycode) {
+    if (
+        ((keycode & (0b1100 << 12)) != (0b1000 << 12))  &&  // not ACT_LAYER, ACT_LAYER_TAP
+        ((keycode & (0b1111 << 12)) != (0b0110 << 12))  &&  // not ACT_SWAP_HANDS
+        ((keycode & (0b1100 << 12)) != (0b0000 << 12))      // not ACT_MODS, ACT_MODS_TAP
```
Comment (sigprof) [inline quantum/keycode_config.c:30]: This is not correct — apparently you are using the `action_t` bits here, but this code handles quantum keycodes, therefore you should check for the appropriate keycode ranges instead:

- `QK_BASIC ... QK_MODS_MAX` (this covers both basic keycodes and keycodes with extra modifiers);
- `QK_LAYER_TAP ... QK_LAYER_TAP_MAX`;
- `QK_SWAP_HANDS ... QK_SWAP_HANDS_MAX` (in this case the low byte might also contain some special values, but these values should not be changed by the magic settings anyway);
- `QK_MOD_TAP ... QK_MOD_TAP_MAX`.

**Review by sigprof — CHANGES_REQUESTED** [review]
Did you look at another implementation of basically the same thing in #9126?

One thing that looks strange in the resulting API is that `keycode_config()` remaps the embedded basic keycodes in the passed value, but does not remap the embedded mod bits in some of those keycodes, therefore the conversion performed by this function ends up incomplete. Maybe a more consistent API would include that conversion too (e.g., rename the existing `keycode_config()` to `uint8_t basic_keycode_config(uint8_t keycode)`, and then implement `uint16_t keycode_config(uint16_t keycode)` in terms of `basic_keycode_config()` and `mod_config()`; also remove the `mod_config()` calls from `action_for_keycode()`). Although I'm not 100% sure about this part.

**Comment by osamuaoki** [comment]
Thanks for the correction.  Yah, that's makes sense.

**Comment by osamuaoki** [comment]
I don't understand why "bootmagic config" is used to describe "magic" in the original source code.  I don't see this expression elsewhere.  My new comment followed sigprof's text thus use "magic"

**Comment by osamuaoki** [comment]
Hmmm... I now guess why "boot" in "bootmagic".  Magic is persistent and boot time default.  It has little or nothing in common with "BOOTMAGIC LITE" ESC-power-on programming.  Still confusing.

**Comment by drashna** [comment]
> I don't understand why "bootmagic config" is used to describe "magic" in the original source code. I don't see this expression elsewhere. My new comment followed sigprof's text thus use "magic"

Because bootmagic also controlled a lot of this functionality, too.  However, because it is too easy to accidentally trigger, we removed the bootmagic code to do this completely. (a LOT of tickets of "why are these keys switched?!?!")

**Comment by osamuaoki** [comment]
> > I don't understand why "bootmagic config" is used to describe "magic" in the original source code. I don't see this expression elsewhere. My new comment followed sigprof's text thus use "magic"
> 
> Because bootmagic also controlled a lot of this functionality, too. However, because it is too easy to accidentally trigger, we removed the bootmagic code to do this completely. (a LOT of tickets of "why are these keys switched?!?!")

I see. But if so, why we use persistent memory to store magic.  For me, MAGIC_SWAP_FOO feature is useful as a temporary work around  when I plugged in to a machine with some special input handling.  We can program QMK default map for our persistent preference anyway if you programmed it.  For end user products, they probably add DIP switch to set it.

I now see this persistent variables for magic is defined in bootmagic.  Why not making this keymap_config struct as a simple static variable.  This is easily done by adding a compiler flag with a few lines of code so keeping backward compatibility option is easily implemented.  So no one get confused if you accidentaly change magic swap setting.  Re-plug keyboard will always reset magic swap settings.

**Comment by osamuaoki** [comment]
Hi,

Applied extra patch to address possible issues:
https://github.com/qmk/qmk_firmware/pull/9126#issuecomment-1247576810

I think all these can be squashed

**Comment by sigprof** [comment]
The standard practice in QMK is to squash commits when merging a PR, so you don't need to worry about that — use separate commits if it's easier for you.

File: quantum/keycode_config.c (hunk 1 above)
Comment (osamuaoki) [reply quantum/keycode_config.c]: This has been addressed

**Pull request closed without merging**